[18.0][MIG] Account_move_line_reconcile_manual: Migration to 18.0

TB-Ph35 opened the pull request.

alexis-via committed:
[ADD] module account_move_line_reconcile_manual

legalsylvain committed:
[FIX] documentation : images must be preceded by a line break

alexis-via committed:
account_move_line_reconcile_manual: add tests

alexis-via committed:
reconcile_manual: add analytic validation when posting write-off

alexis-via committed:
reconcile_manual: set default analytic distribution if model matches criteria

Commit:
[UPD] Update account_move_line_reconcile_manual.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[ADD] icon.png

Gelo-fl committed:
Added translation using Weblate (Spanish)

Gelo-fl committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (50 of 50 strings)

Translation: account-reconcile-16.0/account-reconcile-16.0-account_move_line_reconcile_manual
Translate-URL: https://translation.odoo-community.org/projects/account-reconcile-16-0/account-reconcile-16-0-account_move_line_reconcile_manual/es/

Commit:
[UPD] Update account_move_line_reconcile_manual.pot

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-reconcile-16.0/account-reconcile-16.0-account_move_line_reconcile_manual
Translate-URL: https://translation.odoo-community.org/projects/account-reconcile-16-0/account-reconcile-16-0-account_move_line_reconcile_manual/

badbole committed:
Added translation using Weblate (Croatian)

badbole committed:
Translated using Weblate (Croatian)

Currently translated at 96.0% (49 of 51 strings)

Translation: account-reconcile-16.0/account-reconcile-16.0-account_move_line_reconcile_manual
Translate-URL: https://translation.odoo-community.org/projects/account-reconcile-16-0/account-reconcile-16-0-account_move_line_reconcile_manual/hr/

florian-dacosta committed:
[IMP] account_move_line_reconcile_manual : Manage reconciliation of line sharing a foreign currency

adrianojprado committed:
Added translation using Weblate (Portuguese (Brazil))

adrianojprado committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 86.2% (44 of 51 strings)

Translation: account-reconcile-16.0/account-reconcile-16.0-account_move_line_reconcile_manual
Translate-URL: https://translation.odoo-community.org/projects/account-reconcile-16-0/account-reconcile-16-0-account_move_line_reconcile_manual/pt_BR/

adrianojprado committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (51 of 51 strings)

Translation: account-reconcile-16.0/account-reconcile-16.0-account_move_line_reconcile_manual
Translate-URL: https://translation.odoo-community.org/projects/account-reconcile-16-0/account-reconcile-16-0-account_move_line_reconcile_manual/pt_BR/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (51 of 51 strings)

Translation: account-reconcile-16.0/account-reconcile-16.0-account_move_line_reconcile_manual
Translate-URL: https://translation.odoo-community.org/projects/account-reconcile-16-0/account-reconcile-16-0-account_move_line_reconcile_manual/es/

remi-filament committed:
Added translation using Weblate (French)

remi-filament committed:
Translated using Weblate (French)

Currently translated at 100.0% (51 of 51 strings)

Translation: account-reconcile-16.0/account-reconcile-16.0-account_move_line_reconcile_manual
Translate-URL: https://translation.odoo-community.org/projects/account-reconcile-16-0/account-reconcile-16-0-account_move_line_reconcile_manual/fr/

OCA-git-bot committed:
[UPD] README.rst

alexis-via committed:
reconcile_manual: add write-off models

Commit:
[UPD] Update account_move_line_reconcile_manual.pot

OCA-git-bot committed:
[BOT] post-merge updates

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-reconcile-16.0/account-reconcile-16.0-account_move_line_reconcile_manual
Translate-URL: https://translation.odoo-community.org/projects/account-reconcile-16-0/account-reconcile-16-0-account_move_line_reconcile_manual/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (61 of 61 strings)

Translation: account-reconcile-16.0/account-reconcile-16.0-account_move_line_reconcile_manual
Translate-URL: https://translation.odoo-community.org/projects/account-reconcile-16-0/account-reconcile-16-0-account_move_line_reconcile_manual/es/

adrianojprado committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (61 of 61 strings)

Translation: account-reconcile-16.0/account-reconcile-16.0-account_move_line_reconcile_manual
Translate-URL: https://translation.odoo-community.org/projects/account-reconcile-16-0/account-reconcile-16-0-account_move_line_reconcile_manual/pt_BR/

Commit:
Translated using Weblate (French)

Currently translated at 100.0% (61 of 61 strings)

Translation: account-reconcile-16.0/account-reconcile-16.0-account_move_line_reconcile_manual
Translate-URL: https://translation.odoo-community.org/projects/account-reconcile-16-0/account-reconcile-16-0-account_move_line_reconcile_manual/fr/

mymage committed:
Added translation using Weblate (Italian)

jakobkrabbe committed:
Added translation using Weblate (Swedish)

jakobkrabbe committed:
Translated using Weblate (Swedish)

Currently translated at 100.0% (61 of 61 strings)

Translation: account-reconcile-16.0/account-reconcile-16.0-account_move_line_reconcile_manual
Translate-URL: https://translation.odoo-community.org/projects/account-reconcile-16-0/account-reconcile-16-0-account_move_line_reconcile_manual/sv/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (61 of 61 strings)

Translation: account-reconcile-16.0/account-reconcile-16.0-account_move_line_reconcile_manual
Translate-URL: https://translation.odoo-community.org/projects/account-reconcile-16-0/account-reconcile-16-0-account_move_line_reconcile_manual/it/

florian-dacosta committed:
account_move_line_reconcile_manual : Add Roadmap with known issue

jakobkrabbe committed:
Translated using Weblate (Swedish)

Currently translated at 100.0% (61 of 61 strings)

Translation: account-reconcile-16.0/account-reconcile-16.0-account_move_line_reconcile_manual
Translate-URL: https://translation.odoo-community.org/projects/account-reconcile-16-0/account-reconcile-16-0-account_move_line_reconcile_manual/sv/

peterromao committed:
Added translation using Weblate (Portuguese)

peterromao committed:
Translated using Weblate (Portuguese)

Currently translated at 100.0% (61 of 61 strings)

Translation: account-reconcile-16.0/account-reconcile-16.0-account_move_line_reconcile_manual
Translate-URL: https://translation.odoo-community.org/projects/account-reconcile-16-0/account-reconcile-16-0-account_move_line_reconcile_manual/pt/

SottomaiorMacedoTec committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (61 of 61 strings)

Translation: account-reconcile-16.0/account-reconcile-16.0-account_move_line_reconcile_manual
Translate-URL: https://translation.odoo-community.org/projects/account-reconcile-16-0/account-reconcile-16-0-account_move_line_reconcile_manual/pt_BR/

Commit:
[IMP] account_move_line_reconcile_manual: pre-commit execution

Commit:
[MIG] account_move_line_reconcile_manual: Migration to 18.0

Commit:
[MIG] Account_move_line_reconcile_manual

Commit:
FIX pre-commit

Commit:
[FIX] field company_currency_id in .po

alexis-via committed:
[FIX] account_move_line_reconcile_manual: fix multi-currency full reconcile and reconcile with write-off

florian-dacosta committed:
[IMP][account_move_line_reconcile_manual] Improve foreign currency tests

florian-dacosta committed:
[FIX] account_move_line_reconcile_manual: mono foreign currency reconciliation needs the exchange difference move

florian-dacosta committed:
[FIX] company in case of shared accounts and visibility of write-off option

florian-dacosta commented:
Waiting to decide on https://github.com/akretion/account-reconcile/pull/5
Port of https://github.com/OCA/account-reconcile/pull/934

Reyes4711-S73 commented on account_move_line_reconcile_manual/models/account_reconcile_manual_model.py:
```diff
@@ -0,0 +1,99 @@
+                    "company_id": model.company_id.id,
+                }
+            )
+            model.expense_analytic_distribution = expense_distri
+            model.income_analytic_distribution = income_distri
```
```suggestion
            model.update(
                {
                    "expense_analytic_distribution": expense_distri,
                    "income_analytic_distribution": income_distri,
                }
            )
```

florian-dacosta committed:
Merge pull request #5 from akretion/18-no-exchange-diff-multiple-currency

18 no exchange diff multiple currency

florian-dacosta committed:
[REF] Allow to complete reconciliation on partially reconciled entries
It changes a little bit the process, you cant select reconciled lines anymore (lines having a null amount residual but a belonging to a partial reconciliation)
If really needed you need to un-reconcile the line first. But the right way to do it is to select only lines with a not null amount_residual (even if partially reconciled)

florian-dacosta committed:
Merge pull request #6 from akretion/18-account_move_line_reconcile_manual-partial-management

[REF] Allow to complete reconciliation on partially reconciled entries

jbaudoux reviewed (commented):
@florian-dacosta If I have an invoice in foreign currency and a payment in company currency, I can't reconcile with write-off. The created write-off line makes the residual amount to 0 but the invoice currency residual amount is not 0 and the currency line is not considered as reconciled.

BhaveshHeliconia commented on account_move_line_reconcile_manual/models/account_reconcile_manual_model.py:
```diff
@@ -0,0 +1,99 @@
+            res["journal_id"] = journals.id
+        return res
+
+    @api.depends("expense_account_id", "income_account_id")
```
```suggestion
    @api.depends("expense_account_id", "income_account_id")
    def _compute_analytic_distribution(self):
        aadmo = self.env["account.analytic.distribution.model"]
        for model in self:
            expense_code = model.expense_account_id.code
            income_code = model.income_account_id.code

            model.expense_analytic_distribution = aadmo._get_distribution({
                "account_prefix": expense_code,
                "company_id": model.company_id.id,
            }) if expense_code else {}

            model.income_analytic_distribution = aadmo._get_distribution({
                "account_prefix": income_code,
                "company_id": model.company_id.id,
            }) if income_code else {}
```

florian-dacosta commented:
> @florian-dacosta If I have an invoice in foreign currency and a payment in company currency, I can't reconcile with write-off. The created write-off line makes the residual amount to 0 but the invoice currency residual amount is not 0 and the currency line is not considered as reconciled.

@jbaudoux 
I may try to check this case, but these multi-currency reconciliation cases are not always well managed by Odoo (at least it was not at all in v16). 
We manage well the "same currency" case (company currency or foreign), but multi company currency cases are not necessarily working indeed. I'd need to take the time to check if Odoo improved it in this version allowing us to have something more reliable about these cases.

florian-dacosta committed:
[IMP] Add simple test about multiple currency reconciliation

florian-dacosta commented:
@jbaudoux Like I said, before (v16) multiple currency reconciliation were quite buggy, not really supported by the module (see https://github.com/OCA/account-reconcile/pull/558)

But I just added a simple multi-currency test (that was not working in v16) and it is passing now.
Also I did 2/3 manual simple test with invoice in foreign currency, payment in company currency + write-off, and it did work.
Could you describe the steps to reproduce your use case exactly ?

jbaudoux commented:
> @jbaudoux Like I said, before (v16) multiple currency reconciliation were quite buggy, not really supported by the module (see #558)
> 
> But I just added a simple multi-currency test (that was not working in v16) and it is passing now. Also I did 2/3 manual simple test with invoice in foreign currency, payment in company currency + write-off, and it did work. Could you describe the steps to reproduce your use case exactly ?

Thanks for looking into this. You could complete your test by appending some extra checks:
Line1 residual amount is 0
Line1 currency residual amount is 0
Line1 reconciled is True

In your manual test, have you changed the currency rate before reconciliation so that the currency conversion of the write off doesn't match the currency residual?

florian-dacosta committed:
[IMP] account_move_line_reconcile_manual: refactore tests to be able to test multiple  scenarios easily

florian-dacosta committed:
[REF] Manage multi-currency reconciliation
In case of multiple debit/credit lines to reconcile, with multiple currencies (company currency + foreign currency), it is very hard to guess what what amounts should be the writeoff to complete the reconciliaton.
The reconciliation is always done between one debit and one credit line. Depending on the currency, or even the account of the debit and credit line, Odoo may compare foreign currency amount, company currency amount, with different rates...and it may be different at each iteration.
For these cases, instead of guessing the writeoff amounts, we do simulate the reconciliation to have the real needed amounts.

florian-dacosta commented:
@jbaudoux I did refactore the tests to be able to add different cases more easily.
And adding some differente rates did show issues on reconciliation between foreign currency and company currency indeed.

Like I said, these cases are quite complicated, but in v18, I think there is a way to get around the issue.
I did a refactore here : https://github.com/akretion/account-reconcile/pull/7
I am interested to know if you still have issues with this one.

florian-dacosta committed:
Update account_move_line_reconcile_manual/wizards/account_move_line_reconcile_manual.py

Co-authored-by: Jacques-Etienne Baudoux <je@bcim.be>

florian-dacosta committed:
Merge pull request #7 from akretion/18-manual-ref-multi-ucrr

[18][REF] account_move_line_reconcile_manual: Manage multiple currency reconciliation
